src/flag_gems/ops/dropout.py

```diff
@@ -59,101 +60,64 @@ def dropout_forward_kernel(
-    off_1 = off_0 + BLOCK
-    off_2 = off_1 + BLOCK
-    off_3 = off_2 + BLOCK
-
-    dy_0 = tl.load(DY + off_0, mask=off_0 < N, other=0.0, eviction_policy="evict_first")
-    dy_1 = tl.load(DY + off_1, mask=off_1 < N, other=0.0, eviction_policy="evict_first")
-    dy_2 = tl.load(DY + off_2, mask=off_2 < N, other=0.0, eviction_policy="evict_first")
-    dy_3 = tl.load(DY + off_3, mask=off_3 < N, other=0.0, eviction_policy="evict_first")
-
-    p = 1.0 / (1.0 - p)
-    dx_0 = p * dy_0 * mask0
-    dx_1 = p * dy_1 * mask1
-    dx_2 = p * dy_2 * mask2
-    dx_3 = p * dy_3 * mask3
-
-    tl.store(DX + off_0, dx_0, mask=off_0 < N, eviction_policy="evict_first")
-    tl.store(DX + off_1, dx_1, mask=off_1 < N, eviction_policy="evict_first")
-    tl.store(DX + off_2, dx_2, mask=off_2 < N, eviction_policy="evict_first")
-    tl.store(DX + off_3, dx_3, mask=off_3 < N, eviction_policy="evict_first")
+    offset = tl.program_id(0) * BLOCK + tl.arange(0, BLOCK)
+    mask = offset < N
+    m = tl.load(M + offset, mask=mask, other=0, eviction_policy="evict_first")
+    dy = tl.load(DY + offset, mask=mask, other=0, eviction_policy="evict_first")
+    dx = dy * m * scale
+    tl.store(DX + offset, dx, mask=mask, eviction_policy="evict_first")
 
 
 UNROLL = 4
 
 
-class NativeDropout(torch.autograd.Function):
-    @staticmethod
-    def forward(ctx, x, p, train):
-        logging.debug("GEMS NATIVE DROPOUT FORWARD")
-        assert p > 0.0 and p < 1.0, "p must be in (0, 1)"
-        device = x.device
-        x = x.contiguous()
-        out = torch.empty_like(x)
-        N = x.numel()
-        grid_fn = lambda meta: (triton.cdiv(N, meta["BLOCK"] * UNROLL),)
-        # (TODO) Using Triton autotuner makes kernel parameters opaque to the caller,
-        # hence we cannot obtain the per thread offset as in Pytorch.
-        increment = triton.cdiv(N, UNROLL)
-        with torch_device_fn.device(device):
-            philox_seed, philox_offset = philox_backend_seed_offset(increment)
-            dropout_forward_kernel[grid_fn](x, out, N, p, philox_seed, philox_offset)
-        ctx.p = p
-        ctx.philox_seed = philox_seed
-        ctx.philox_offset = philox_offset
-        return out, None
-
-    @staticmethod
-    def backward(ctx, grad_outputs, kwargs):
-        logging.debug("GEMS NATIVE DROPOUT BACKWARD")
-        device = grad_outputs.device
-        grad_outputs = grad_outputs.contiguous()
-        grad_inputs = torch.empty_like(grad_outputs)
-        N = grad_outputs.numel()
-        grid_fn = lambda meta: (triton.cdiv(N, meta["BLOCK"] * UNROLL),)
-        with torch_device_fn.device(device):
-            dropout_backward_kernel[grid_fn](
-                grad_outputs, grad_inputs, N, ctx.p, ctx.philox_seed, ctx.philox_offset
-            )
-        return grad_inputs, None, None
-
-
-def native_dropout(x, p=0.5, train=True):
-    return NativeDropout.apply(x, p, train)
+def dropout(input, p, train):
```

Comment:
I realized we didn't handle we train=False correctly in the previous version. Let's fix that.

Reply:
done.